### PR TITLE
feat: fetch link titles for urls in message text

### LINKED ISSUE #43: feat: extract title and meta from urls of chat messages

### PR DESCRIPTION
**Pull request by OverflowCat**
Close #43

### CODE REVIEW

File: pkg/bots/tgbot/utils.go [hunk 1]
```diff
@@ -82,11 +86,49 @@ func FullNameFromFirstAndLastName(firstName, lastName string) string {
 }
 
 func ExtractTextFromMessage(message *tgbotapi.Message) string {
-	if message.Caption != "" {
-		return message.Caption
+	text := lo.Ternary(message.Caption != "", message.Caption, message.Text)
+	type MarkdownLink struct {
```
Comment (nekomeowww) [inline pkg/bots/tgbot/utils.go:90]:
```suggestion

	type MarkdownLink struct {
```

File: pkg/bots/tgbot/utils.go [hunk 2]
```diff
@@ -82,11 +86,49 @@ func FullNameFromFirstAndLastName(firstName, lastName string) string {
...
+		Markdown []uint16
+		Start    int
+		End      int
 	}
 
-	return message.Text
+	text_utf16 := utf16.Encode([]rune(text))
```
Comment (nekomeowww) [inline pkg/bots/tgbot/utils.go:96]: snake_case is not idiomatic in Go, should change to camelCase

File: pkg/bots/tgbot/utils_test.go [hunk 3]
```diff
@@ -4,12 +4,12 @@ import (
...
 	"testing"
 
+	tgbotapi "github.com/go-telegram-bot-api/telegram-bot-api/v5"
 	"github.com/stretchr/testify/assert"
 )
 
 func TestReplaceMarkdownTitlesToBoldTexts(t *testing.T) {
 	prefix := ""
-
 	for i := 0; i < 6; i++ {
```
Comment (nekomeowww) [inline pkg/bots/tgbot/utils_test.go:13]:
```suggestion

	for i := 0; i < 6; i++ {
```

File: pkg/bots/tgbot/utils.go [hunk 4]
```diff
@@ -82,11 +86,51 @@ func FullNameFromFirstAndLastName(firstName, lastName string) string {
...
+
+	textUtf16 := utf16.Encode([]rune(text))
+	links := lop.Map(message.Entities, func(entity tgbotapi.MessageEntity, i int) MarkdownLink {
+		startIndex := entity.Offset
+		endIndex := startIndex + entity.Length
+		var title string
+		var href string
+		if entity.Type == "url" {
+			href = string(utf16.Decode(textUtf16[startIndex:endIndex]))
+			result, err := LinkPreview.PreviewLink(href, nil)
```
Comment (nekomeowww) [inline pkg/bots/tgbot/utils.go:105]: This err should be logged, but this err doesnt really matter... How about adding an err field to MarkdownLink to handle the outside functions?

File: pkg/bots/tgbot/utils.go [hunk 5]
```diff
@@ -82,11 +86,51 @@ func FullNameFromFirstAndLastName(firstName, lastName string) string {
...
-		return message.Caption
+	text := lo.Ternary(message.Caption != "", message.Caption, message.Text)
+
+	type MarkdownLink struct {
+		Markdown []uint16
+		Start    int
+		End      int
+	}
+
+	textUtf16 := utf16.Encode([]rune(text))
```
Comment (nekomeowww) [inline pkg/bots/tgbot/utils.go:97]: According to the Go naming convention, Utf should be UTF.

File: pkg/bots/tgbot/utils.go [hunk 6]
```diff
@@ -82,11 +86,51 @@ func FullNameFromFirstAndLastName(firstName, lastName string) string {
...
+			}
+			title = result.Title
+		} else if entity.Type == "text_link" {
+			title = string(utf16.Decode(textUtf16[startIndex:endIndex]))
+			href = entity.URL
+		} else {
+			return MarkdownLink{[]uint16{}, -1, -1}
+		}
+		unescaped, err := url.QueryUnescape(href)
+		if err == nil {
```
Comment (nekomeowww) [inline pkg/bots/tgbot/utils.go:117]: What about err != nil?

File: pkg/bots/tgbot/utils.go (hunk 4 above)
Comment (OverflowCat) [reply pkg/bots/tgbot/utils.go:105]: Many sites does not have a preview (say, Twitter) and doing so may make the code hard to maintain. Maybe we can move this function to `chat_histories.go`.

Comment (nekomeowww) [reply pkg/bots/tgbot/utils.go:105]: Sure.

**Comment by nekomeowww** [comment]
Could you please update the branch and rebase with `upstream/main`?

File: pkg/bots/tgbot/utils.go (hunk 6 above)
Comment (OverflowCat) [reply pkg/bots/tgbot/utils.go:117]: Unescaping the link is to make it interpretable for AI as well as save tokens. Note that the link is not necessarily a URI encoded one, and errors occur when `%` has no succeeding hex value. A link may contain many `%` without being encoded. And if it is unescapable, we just keep it. Or, is it a convention to write `if err != nil { } else { // do something }`?

Comment (nekomeowww) [reply pkg/bots/tgbot/utils.go:117]: Nah, it is ok to omit the err.

But I have checked the source code of `url.QueryEscape(...)` and find out that `unescaped` is always empty if any err occurred, so you might have to deal with the empty string I guess?

File: go.mod [hunk 7]
```diff
@@ -4,6 +4,7 @@ go 1.19
 
 require (
 	entgo.io/ent v0.12.2
+	github.com/Junzki/link-preview v0.0.0-20190616153136-4d94cb3f9ebf
```
Comment (nekomeowww) [inline go.mod:7]: This package had a very ugly implementation and didn't quite follow the regulations and best practices in Go. 

However, I think such logic is simple enough to wrap one by ourselves. I will make one for you.

Comment (OverflowCat) [reply go.mod:7]: Didn't find any well-maintained packages, though.

File: pkg/bots/tgbot/utils.go (hunk 6 above)
Comment (OverflowCat) [reply pkg/bots/tgbot/utils.go]: `if err != nil` then `href` won't be modified and href is kept as it is.

File: go.mod (hunk 7 above)
Comment (nekomeowww) [reply go.mod:7]: It's ok. Thank you for doing the research!

File: pkg/bots/tgbot/utils.go (hunk 6 above)
Comment (nekomeowww) [reply pkg/bots/tgbot/utils.go]: Understood.

**Review by nekomeowww — DISMISSED** [review]

**Review by nekomeowww — APPROVED** [review]

**Pull request merged**